lib/Backend/EhFrame.cpp

```diff
@@ -0,0 +1,226 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#include "Backend.h"
+#include "EhFrame.h"
+
+// AMD64 ABI -- DWARF register number mapping
+static const ubyte DWARF_RegNum[] =
+{
+    // Exactly same order as RegList.h!
+    -1, // NOREG,
+    0,  // RAX,
+    2,  // RCX,
+    1,  // RDX,
+    3,  // RBX,
+    7,  // RSP,
+    6,  // RBP,
+    4,  // RSI,
+    5,  // RDI,
+    8,  // R8,
+    9,  // R9,
+    10, // R10,
+    11, // R11,
+    12, // R12,
+    13, // R13,
+    14, // R14,
+    15, // R15,
+    17,  // XMM0,
+    18,  // XMM1,
+    19,  // XMM2,
+    20,  // XMM3,
+    21,  // XMM4,
+    22,  // XMM5,
+    23,  // XMM6,
+    24,  // XMM7,
+    25,  // XMM8,
+    26,  // XMM9,
+    27,  // XMM10,
+    28,  // XMM11,
+    29,  // XMM12,
+    30,  // XMM13,
+    31,  // XMM14,
+    32,  // XMM15,
+};
+
+static const ubyte DWARF_RegRA = 16;
+
+ubyte GetDwarfRegNum(ubyte regNum)
+{
+    return DWARF_RegNum[regNum];
+}
+
+// Enocde into ULEB128 (Unsigned Little Endian Base 128)
+BYTE* EmitLEB128(BYTE* pc, unsigned value)
+{
+    do
+    {
+        BYTE b = value & 0x7F; // low order 7 bits
+        value >>= 7;
+
+        if (value)  // more bytes to come
+        {
+            b |= 0x80;
+        }
+
+        *pc++ = b;
+    }
+    while (value != 0);
+
+    return pc;
+}
+
+// Encode into signed LEB128 (Signed Little Endian Base 128)
+BYTE* EmitLEB128(BYTE* pc, int value)
+{
+    static const int size = sizeof(value) * 8;
+    static const bool isLogicShift = (-1 >> 1) != -1;
+
+    const bool signExtend = isLogicShift && value < 0;
+
+    bool more = true;
+    while (more)
+    {
+        BYTE b = value & 0x7F; // low order 7 bits
+        value >>= 7;
+
+        if (signExtend)
+        {
+            value |= - (1 << (size - 7)); // sign extend
+        }
+
+        const bool signBit = (b & 0x40) != 0;
+        if ((value == 0 && !signBit) || (value == -1 && signBit))
+        {
+            more = false;
+        }
+        else
+        {
+            b |= 0x80;
+        }
+
+        *pc++ = b;
+    }
+
+    return pc;
+}
+
+
+void EhFrame::Entry::Begin()
+{
+    Assert(beginOffset == -1);
+    beginOffset = writer->Count();
+
+    // Write Length place holder
+    const uword length = 0;
+    writer->Write(length);
+}
+
+void EhFrame::Entry::End()
+{
+    // padding
+    size_t padding = (MachPtr - writer->Count() % MachPtr) % MachPtr;
+    for (size_t i = 0; i < padding; i++)
+    {
+        cfi_nop();
+    }
+
+    // update length record
+    uword length = writer->Count() - beginOffset
```

Comment:
I don't see this assert to be very useful. Once Begin() was called this would hardly fail as the same uword type is written again at beginOffset. Maybe I can assert beginOffset != -1 to ensure Begin() was called?